### DISCUSSION TITLE
Filter duplicate favorites on import

### DISCUSSION THREAD

**Discussion by Ciutnik** [devemux86/cruiser#491] [Q&A] [answered]
I know it's not directly related to Cruiser's core functionality, but since we have the great Overpass-turbo mentioned here: https://github.com/devemux86/cruiser/discussions/488#discussioncomment-9080369, maybe it wouldn't be complicated to implement filtering of duplicates (in the same location) POIs during GPX import?

What I mean is that by running a query every so often, and importing the resulting GPX file, only NEW points would be added.

~~Or do you know of any tool for this? I haven't found anything. Even GPS-Babel didn't help me....~~

EDIT: Strange, now I checked and GPS Babel seems to do it correctly. I must have gotten it wrong previously (I was using the command line).

**Comment by devemux86** [comment]
Do you mean filter 2 gpx files and keep the unique points?
Unique by what value? Name or coordinates?

**Reply by Ciutnik** [reply]
The minimum option would be to check when importing GPX into a group whether a point in the ~same LOCATION already exists. And the ideal option would be to check the entire collection for duplicates. I'm still thinking about the coordinates (radius?) because names can naturally repeat.

**Reply by devemux86** [reply]
There is already a check in the favorites when they are imported in an existing group.
The check compares their coordinates and name. Both must be "equal" to filter a "same" favorite.

**Reply by Ciutnik** [reply]
This is good news! Great 👍 If the name is different but the coordinates are the same then a duplicate will be created? I can deal with this by carefully watching the names 😉

And what happens when the imported, source file already contains duplicates?

I often rename POIs imported from Overpass, because those that do not have a name of their own in the OSM tag show up as, for example, `way/1234567890` or `node/1234567890`. So, in this case I batch rename them to the same name as the Group has.

**Reply by devemux86** [reply]
> If the name is different but the coordinates are the same then a duplicate will be created?

It is not sure if this is indeed a duplicate or simply another favorite at the same location.
So different names mean different favorites, regardless the location.

> And what happens when the imported, source file already contains duplicates?

The filtering happens on the imported favorites versus the existing favorites.
If an imported gpx has already "duplicate" favorites, currently they are all used.

**Reply by Ciutnik** [reply]
> If an imported gpx has already "duplicate" favorites, currently they are all used.

This is already worse news, but I understand and could expect it. The most important thing is that new duplicates will not be created!

Indeed, in the case of a different name and the same coordinates may be two different points. Yet by design they should be in different groups.... but they don't have to 😊

**Comment by MotoUKRider** [comment]
Running v-b3 here, and the duplicate check fails when importing a ZIP of GPX files ([as reported here](https://github.com/devemux86/cruiser/discussions/495#discussioncomment-9192571)).  The Name+Lat/Lon are the same, with identical decimal places.

I use this method to slipstream a handful of new Favorites into the Groups, and from the comment above, Cruiser *should* keep the Groups intact rather than create a duplicate set.  As things stand, I have to swipe away my existing groups before a fresh import.

PS: I see you've added `x:visible` which I've added to my GPX maker, as I was nearly going to report a "no visibility" bug :fearful:

**Reply by devemux86** [reply]
Do you mean check for same group names?

Check in favorites is already done when import them in a group.

**Reply by MotoUKRider** [reply]
@devemux86 Yes, I suppose that is the cause, Cruiser sees "Ports" and then adds "Ports 2" so the check doesn't happen.

**Reply by devemux86** [reply]
@MotoUKRider this is the intended behavior, because it was asked in the past to create group 2, group 3, etc.

But we can change it if we want it.

**Reply by MotoUKRider** [reply]
@devemux86 If it's an easy fix that pleases both parties, perhaps.
The issue is that I end up with 2 entries for each Favourite icon.  I'm not used to duplicates, BaseCamp won't allow them!

The New Group/Update Existing Group option would be useful when loading a ZIP.  At present, I have 9 groups in the ZIP - the swiping isn't a major hassle, but the process didn't seem right with the dupes being created.  If users eventually have groups for *all* the icons you offer, and use a ZIP to populate them - it could be a problem.

I see that things are OK when I import each GPX on its own - but doing that 9x *is* a bit of a hassle :shrug:

**Reply by devemux86** [reply]
Group names cannot be duplicates, it is not allowed by the file system.
Instead of creating groups 2, 3 we could merge the imports into existing groups with the same name?
I will check it.

**Reply by devemux86** [reply]
@MotoUKRider 

Import groups of favorites / routes / tracks was improved: update groups with same name.
You can all test it in the next Cruiser (Beta) update.

**Reply by MotoUKRider** [reply]
@devemux86 Thanks for the fix - the Group import works nicely after a few tests.

**Reply by devemux86** [reply]
@MotoUKRider 

There are improvements in all 3 types: favorites + routes + tracks.
Import and export: groups and single or multiple selected items.

**Comment by devemux86** [comment, marked as answer]
Implemented in [Cruiser 4.0.1](https://play.google.com/store/apps/details?id=gr.talent.cruiser).
